subsys/bluetooth/services/dis.c

```diff
@@ -45,105 +48,269 @@ static struct dis_pnp dis_pnp_id = {
+#define BT_DIS_SW_REV_STR_REF        CONFIG_BT_DIS_SW_REV_STR
+#define BT_DIS_UDI_LABEL_STR_REF     CONFIG_BT_DIS_UDI_LABEL_STR
+#define BT_DIS_UDI_DI_STR_REF        CONFIG_BT_DIS_UDI_DI_STR
+#define BT_DIS_UDI_ISSUER_STR_REF    CONFIG_BT_DIS_UDI_ISSUER_STR
+#define BT_DIS_UDI_AUTHORITY_STR_REF CONFIG_BT_DIS_UDI_AUTHORITY_STR
+#define BT_DIS_IEEE_RCDL_STR_REF     CONFIG_BT_DIS_IEEE_RCDL_STR
 
 #endif /* CONFIG_BT_DIS_SETTINGS */
 
-static ssize_t read_str(struct bt_conn *conn,
-			  const struct bt_gatt_attr *attr, void *buf,
-			  uint16_t len, uint16_t offset)
+static ssize_t read_str(struct bt_conn *conn, const struct bt_gatt_attr *attr, void *buf,
+			uint16_t len, uint16_t offset)
 {
 	return bt_gatt_attr_read(conn, attr, buf, len, offset, attr->user_data,
 				 strlen(attr->user_data));
 }
 
 #if CONFIG_BT_DIS_PNP
-static ssize_t read_pnp_id(struct bt_conn *conn,
-			   const struct bt_gatt_attr *attr, void *buf,
+static ssize_t read_pnp_id(struct bt_conn *conn, const struct bt_gatt_attr *attr, void *buf,
 			   uint16_t len, uint16_t offset)
 {
-	return bt_gatt_attr_read(conn, attr, buf, len, offset, &dis_pnp_id,
-				 sizeof(dis_pnp_id));
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, &dis_pnp_id, sizeof(dis_pnp_id));
+}
+#endif
+
+#if CONFIG_BT_DIS_SYSTEM_ID
+static ssize_t read_system_id(struct bt_conn *conn, const struct bt_gatt_attr *attr, void *buf,
+			      uint16_t len, uint16_t offset)
+{
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, dis_system_id,
+				 sizeof(dis_system_id));
+}
+#endif
+
+#if CONFIG_BT_DIS_UDI
+#define BT_DIS_UDI_FLAG_LABEL     (!!BT_DIS_UDI_LABEL_STR_REF[0])
+#define BT_DIS_UDI_FLAG_DI        (!!BT_DIS_UDI_DI_STR_REF[0])
+#define BT_DIS_UDI_FLAG_ISSUER    (!!BT_DIS_UDI_ISSUER_STR_REF[0])
+#define BT_DIS_UDI_FLAG_AUTHORITY (!!BT_DIS_UDI_AUTHORITY_STR_REF[0])
+#define BT_DIS_UDI_FLAGS                                                                           \
+	(BT_DIS_UDI_FLAG_LABEL | (BT_DIS_UDI_FLAG_DI << 1) | (BT_DIS_UDI_FLAG_ISSUER << 2) |       \
+	 (BT_DIS_UDI_FLAG_AUTHORITY << 3))
+
+/*
+ * UDI for medical devices contains a flag and 4 different null-terminated strings that may have
+ * unknown length. This requires its own encode method.
+ */
+static void read_udi_subval(const char *str, uint16_t str_len, char *buf, uint16_t *bytes_read,
+			    uint16_t *index, uint16_t len, uint16_t offset)
+{
+	/* String should not be with included null-terminator if empty */
+	if (str_len == 0) {
+		return;
+	}
+
+	if (*bytes_read == len) {
+		return;
+	}
+
+	if (*index + str_len + sizeof('\0') < offset) {
+		*index += str_len + sizeof('\0');
+	}
+
+	for (uint16_t i = 0; i < str_len; i++) {
+		if (*index >= offset && *bytes_read < len) {
+			buf[*bytes_read] = str[i];
+
+			(*bytes_read)++;
+		}
+
+		(*index)++;
+	}
+
+	/* Null-terminator */
+	if (*index >= offset && *bytes_read < len) {
+		buf[*bytes_read] = 0;
+
+		(*bytes_read)++;
+	}
```

Comment:
I think this could be simplified a bit if `str_len` was replaced with `val_len` that includes the NULL terminator:

```suggestion
static void read_udi_subval(const char *str, uint16_t val_len, char *buf, uint16_t *bytes_read,
			    uint16_t *index, uint16_t len, uint16_t offset)
{
	/* String should not be with included null-terminator if empty */
	if (val_len == sizeof('\0') {
		return;
	}

	if (*bytes_read == len) {
		return;
	}

	if (*index + val_len < offset) {
		*index += val_len;
	}

	for (uint16_t i = 0; i < val_len; i++) {
		if (*index >= offset && *bytes_read < len) {
			buf[*bytes_read] = str[i];

			(*bytes_read)++;
		}

		(*index)++;
	}
```

I'm also a bit confused about the `index` and the NULL terminator. If we have a string of length 4 (size of 5 with NULL terminator), should the index be increased by 4 or 5? 

In other words, when returning "ABC" + "CDE" as the final value, does that include the NULL terminator between the 2 strings? I assume so, because otherwise it's not possible to split the value into strings after reading.

Reply:
The null-terminators are indeed included in the final string, so index should be incremented by strlen + 1.

Reply:
I made a quick test to gain some confidence that the UDI generation was working at least. See https://gist.github.com/simeneilevstjonn/210ef5db716ea847c2903083fa3dcc4b
Something similar could maybe be in CI.

Reply:
I have now implemented this.

I think the only thing missing now is the System ID thing. I have a suggestion, just waiting for a reply from @alwa-nordic